macOS support for NixOS tests

Gabriella439 opened the pull request:
Closes #193336
Closes #261694
Related to #108984 

This pull request is a composite of the changes from #193336 and also #261694, with some changes of my own that I made.

The goal here was to get the following flake to build and run on `aarch64-darwin`:

```nix
{ inputs.nixpkgs.url = "github:Gabriella439/nixpkgs/gabriella/macOS_NixOS_test";

  outputs = { nixpkgs, ... }: {
    checks.aarch64-darwin.default =
      nixpkgs.legacyPackages.aarch64-darwin.nixosTest {
        name = "test";

        nodes.machine = { };

        testScript = ''
          print(machine.succeed('echo hello'));
        '';
      };
  };
}
```

… and after this change it does.  There's no longer a need for the user to set `nodes.*.nixpkgs.pkgs` or
`nodes.*.virtualisation.host.pkgs` as the correct values are inferred from the host system.

This change is spiritually much closer to the approach in #261694 than the approach in #193336.  However, I still made a few changes compared to #261694:

- I didn't include the change to increase `ulimit`

  I think this change was questionable because I feel like a script provided by Nixpkgs shouldn't be tinkering with the user's ambient `ulimit` settings.

- I named the required system feature `hvf` instead of `apple-virt`

  I preferred the `hvf` system feature name chosen by #193336.

- I didn't use the `node.pkgs` setting to set `nodes.*.nixpkgs.pkgs`

  That does not work, based on my testing, but setting `nixpkgs.system` in `nixos/lib/testing/nodes.nix` does work with some supporting changes to `pkgs/build-support/testers/default.nix`.

## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] aarch64-darwin
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Gabriella439 committed:
macOS support for NixOS tests

Closes #193336
Closes #261694
Related to 108984

This pull request is a composite of the changes from #193336 and
also #261694, with some changes of my own that I made.

The goal here was to get the following flake to build and run on
`aarch64-darwin`:

```nix
{ inputs.nixpkgs.url = <this branch>;

  outputs = { nixpkgs, ... }: {
    checks.aarch64-darwin.default =
      nixpkgs.legacyPackages.aarch64-darwin.nixosTest {
        name = "test";

        nodes.machine = { };

        testScript = "";
      };
  };
}
```

… and after this change it does.  There's no longer a need for the
user to set `nodes.*.nixpkgs.pkgs` or
`nodes.*.virtualisation.host.pkgs` as the correct values are inferred
from the host system.

This change is spiritually much closer to the approach in #261694
than the approach in #193336.  However, I still made a few changes
compared to #261694:

- I didn't include the change to increase `ulimit`

  I think this change was questionable because I feel like a script
  provided by Nixpkgs shouldn't be tinkering with the user's ambient
  `ulimit` settings.

- I named the required system feature `hvf` instead of `apple-virt`

  I preferred the `hvf` system feature name chosen by #193336.

- I didn't use the `node.pkgs` setting to set `nodes.*.nixpkgs.pkgs`

  That does not work, based on my testing, but setting the same thing
  in `nixos/lib/testing/nodes.nix` does work.

- I created a `pkgsLocal` helper

  This is based on the feedback from @roberth here:

  https://github.com/NixOS/nixpkgs/pull/261694/files#r1365443955

  … and is similar in spirit to `pkgsCross`.

cole-h commented:
@ofborg eval

cole-h commented on pkgs/top-level/stage.nix:
```diff
@@ -185,10 +185,25 @@ let
+    # pkgsLocal.x86_64-linux.hello will refer to the "hello" package bult for
+    # x86_64-linux.
+    #
+    # This is essentially the same thing as pkgsCross except for localSystem
+    # instead of crossSystem.
+    pkgsLocal =
```
You'll need to add this new package set to this exclusion list, or ofborg will infinitely recurse: https://github.com/NixOS/nixpkgs/blob/1b439a273dadd4c675e81a0e05abbc9258889c02/pkgs/top-level/release-attrpaths-superset.nix#L50

(We just ran into this issue with https://github.com/NixOS/nixpkgs/pull/274089 :laughing: )

roberth replied:
"Local" isn't really an appropriate term for cross compilation, unfortunately (and we should get rid of it). It is only appropriate in the context of remote builds, but those are not _handled_ by Nixpkgs. Alternatively, "local" may be applicable in the context of a single derivation, but that's also not our context here.

`pkgsStraight` or `pkgsNative` may be better names.

@Ericson2314 any ideas or suggestions for how to name this package set relation?

roberth commented on nixos/lib/testing/macos-host.nix:
```diff
@@ -0,0 +1,18 @@
+
+      # Without this, the store image uses "raw" instead of "qcow2" and
+      # that throws some errors on macOS on startup of the VM
+      # TODO: remove when `raw` errors are fixed.
```
Do we have an issue for this?

roberth commented on nixos/lib/testing/nodes.nix:
```diff
@@ -24,16 +43,18 @@ let
             {
               virtualisation.qemu.package = testModuleArgs.config.qemu.package;
+              virtualisation.host.pkgs = hostPkgs;
+              nixpkgs.pkgs = guestPkgs;
```
We can't "hardcode" a `pkgs` in the default configuration. The default configuration should support all the features of `misc/nixpkgs.nix`, such as `nixpkgs.config`, `nixpkgs.buildPlatform`, etc - ie NixOS needs to call Nixpkgs.

It also happens to break `pkgsReadOnly`:

```nix
nix-repl> testers.runNixOSTest { node.pkgsReadOnly = true; name = "test"; nodes = { machine = { }; }; }
```

This line needs to be moved to a different entrypoint / entrypoints, such as
- `nixosTests`
- `testers.runNixOSTest`

roberth replied:
This is still a problem.

I think we should improve the [node.pkgsReadOnly](https://nixos.org/manual/nixos/stable/#test-opt-node.pkgsReadOnly) option so that its default is not `options.node.pkgs.isDefined` but something like `options.node.pkgs.highestPrio != optionDefaultPrio`. That way the test framework can provide a default guest package set without affecting `node.pkgsReadOnly`'s default.

roberth commented on nixos/lib/testing/nodes.nix:
```diff
@@ -14,6 +14,25 @@ let
+        supportedHosts = lib.concatStringsSep ", " (lib.attrNames hostToGuest);
+
+        message =
+          "Unsupported non-Linux host system: ${hostPlatform.system}, supported: ${supportedHosts}";
```
```suggestion
          "NixOS Test: don't know which VM guest system to pair with VM host system: ${hostPlatform.system}. Perhaps you intended to run the tests on a Linux host, or one of the following systems that may run NixOS tests: ${supportedHosts}";
```

roberth commented on nixos/lib/testing/nodes.nix:
```diff
@@ -24,16 +43,18 @@ let
+          ({ config, pkgs, ... }:
             {
               virtualisation.qemu.package = testModuleArgs.config.qemu.package;
+              virtualisation.host.pkgs = hostPkgs;
```
:+1:

roberth commented on nixos/lib/testing/nodes.nix:
```diff
@@ -24,16 +43,18 @@ let
           ./nixos-test-base.nix
           { key = "nodes"; _module.args.nodes = config.nodesCompat; }
-          ({ config, ... }:
+          ({ config, pkgs, ... }:
```
Unused parameter:

```suggestion
          ({ config, ... }:
```

roberth commented:
> * I named the required system feature `hvf` instead of `apple-virt`

I don't see this in the code.

[`apple-virt`](https://nixos.org/manual/nix/stable/command-ref/conf-file.html?highlight=apple-virt#conf-system-features) is autodetected by Nix 2.19 and should have been backported to 2.18, but that hasn't been released yet.

I've chosen this name because the detection appears to be avf/hvf agnostic.
- See https://github.com/NixOS/nix/pull/9187#issue-1952484105

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/getting-nixos-to-work-macos-m1-2-3-aarch64-darwin/39510/1

Gabriella439 committed:
Improve error message

… as suggested by @roberth

Co-authored-by: Robert Hensing <roberth@users.noreply.github.com>

Gabriella439 committed:
Remove unused `pkgs` parameter

… as suggested by @roberth

Co-authored-by: Robert Hensing <roberth@users.noreply.github.com>

Gabriella439 committed:
s/pkgsLocal/pkgsNative/

… as suggsted by @roberth

Gabriella439 committed:
Exclude `pkgsNative` from build

… as caught by @cole-h

roberth commented on pkgs/top-level/stage.nix:
```diff
@@ -185,10 +185,25 @@ let
 
+    # This maps each entry in lib.systems.examples to its own package set where
+    # packages are built and run on that system.  For example,
+    # pkgsNative.x86_64-linux.hello will refer to the "hello" package bult for
```
```suggestion
    # pkgsNative.x86_64-linux.hello will refer to the "hello" package built for
```

roberth commented on nixos/lib/testing/nodes.nix:
```diff
@@ -27,13 +46,15 @@ let
               mkIf (!options.nixpkgs.pkgs.isDefined) {
                 # TODO: switch to nixpkgs.hostPlatform and make sure containers-imperative test still evaluates.
-                nixpkgs.system = hostPkgs.stdenv.hostPlatform.system;
+                nixpkgs.system = hostPlatform.system;
```
This should probably be the guest system.

Note that this definition only applies when the `pkgs` is not read only - ie we need to construct a new Nixpkgs.
With that in mind, we should not evaluate `pkgsNative.${guestSystem}`, as that Nixpkgs instance will usually not be used.

By improving this definition to use the guest system, we can support darwin as hypervisor for NixOS tests that have overlays, etc.

Gabriella439 commented:
Yeah, I'm planning on getting to fixing the `nixpkgs.pkgs = guestPkgs;`.  I was just puzzling through the best way to do it.

Gabriella439 committed:
Fix typo

… as caught by @robert

Co-authored-by: Robert Hensing <roberth@users.noreply.github.com>

Gabriella439 committed:
Specify guest *system* instead of guest *packages*

… based on feedback from @roberth

This plays nicer with other changes the user might want to make and
also works in conjunction with the `pkgsReadOnly`.

Note that this required deleting the `nixpkgs.pkgs` default because
that's not compatible with setting `nixpkgs.system`.  If `nixpkgs.pkgs`
is defined (even as a default-valued option) then `nixpkgs.system` is
ignored.

Gabriella439 committed:
Revert the addition of `pkgsNative`

This is no longer necessary now that we're overriding `nixpkgs.system`
instead of `nixpkgs.pkgs`

Gabriella439 commented:
@roberth: Alright, I think I finally understood what you were proposing and implemented that.  The main caveat is that I had to get rid of the `nixpkgs.pkgs = lib.mkDefault pkgs;` in the `nixosTest` entry point because that breaks this (and some other stuff, too) since defining `nixpkgs.pkgs` (even just as a default) causes all other `nixpkgs.*` options to be ignored.  I don't know if making that change to the `nixosTest` entry point will break other stuff, though.

Gabriella439 commented on nixos/lib/testing/macos-host.nix:
```diff
@@ -0,0 +1,18 @@
+      # this generally makes rebuilds of the VMs faster and the VMs smaller, while reducing
+      # disk I/O.
+      # However, on macOS it currently leads to many crashing guest processes.
+      virtualisation.useNixStoreImage = true;
```
I actually cannot reproduce this problem on a small test VM, so I'm not sure if I should keep this option (which I cargo-culted from https://github.com/NixOS/nixpkgs/pull/261694)

tfc replied:
Had this on an intel mac. I can try it again on my new apple silicon.

In the mean time, please try again with a big VM. I observed so many random crashes.

Gabriella439 replied:
So I tested this using a variation of the example from [How to use NixOS for lightweight integration testing](https://www.haskellforall.com/2020/11/how-to-use-nixos-for-lightweight.html):

<details>
<summary>Test code</summary>

```nix
{ inputs.nixpkgs.url = "/Users/gabriella/proj/nixpkgs";

  outputs = { nixpkgs, ... }: {
    checks.aarch64-darwin.default =
      let
        database      = "postgres";
        schema        = "api";
        table         = "todos";
        username      = "authenticator";
        password      = "mysecretpassword";
        webRole       = "web_anon";
        postgrestPort = 3000;

      in
        nixpkgs.legacyPackages.aarch64-darwin.nixosTest {
          name = "test";

          nodes = {
            server = { config, pkgs, ... }: {
              environment.defaultPackages = [ pkgs.curl ];

              # Open the default port for `postgrest` in the firewall
              networking.firewall.allowedTCPPorts = [ postgrestPort ];

              services.postgresql = {
                enable = true;

                initialScript = pkgs.writeText "initialScript.sql" ''
                  create schema ${schema};

                  create table ${schema}.${table} (
                    id serial primary key,
                    done boolean not null default false,
                    task text not null,
                    due timestamptz
                  );

                  insert into ${schema}.${table} (task) values
                    ('finish tutorial 0'), ('pat self on back');

                  create role ${webRole} nologin;

                  grant usage on schema ${schema} to ${webRole};
                  grant select on ${schema}.${table} to ${webRole};

                  create role ${username} inherit login password '${password}';
                  grant ${webRole} to ${username};
                '';
              };

              users = {
                mutableUsers = false;

                groups."${username}" = { };

                users = {
                  root.password = "";

                  "${username}" = {
                    isSystemUser = true;
                    group = "${username}";
                  };
                };
              };

              systemd.services.postgrest = {
                wantedBy = [ "multi-user.target" ];

                after = [ "postgresql.service" ];

                script =
                  let
                    configuration = pkgs.writeText "tutorial.conf" ''
                      db-uri = "postgres://${username}:${password}@localhost:${toString config.services.postgresql.port}/${database}"
                      db-schema = "${schema}"
                      db-anon-role = "${username}"
                    '';

                  in
                    ''
                      ${pkgs.postgrest}/bin/postgrest ${configuration}
                    '';

                serviceConfig.User = username;
              };
            };

            client = { };
          };


          testScript =
            ''
            import json
            import sys

            start_all()

            server.wait_for_open_port(${toString postgrestPort})

            expected = [
                {"id": 1, "done": False, "task": "finish tutorial 0", "due": None},
                {"id": 2, "done": False, "task": "pat self on back", "due": None},
            ]

            actual = json.loads(
                client.wait_until_succeeds(
                    "curl --fail http://server:${toString postgrestPort}/${table}"
                )
            )

            if expected != actual:
                sys.exit(1)
            '';
      };
  };
}
```
</details>

… and that test succeeds without issues.  That's running `postgres` and `Postgres`.  If you have a test of your own that was failing before without the workaround you can try running that using this branch and let me know if you're still able to reproduce the problem.  However, currently I'm not able to reproduce the problem.

Gabriella439 committed:
Remove `macos-host.nix`

Based on my local testing these options seem unnecessary and the VM
runs fine without them being set.  I originally cargo-culted these
from #261694 but they don't seem necessary.

Gabriella439 committed:
Remove unused `pkgsNative`

Gabriella439 committed:
Fix evaluation error on Linux

Gabriella439 committed:
Fix test failure

Specifically, the `nixosTest-example` test correctly flagged that the
change from 90679ea06c7f5ad3538d8c3940aba576865609bf broke the user's
expectation that the `pkgs` in `pkgs.nixosTest` would be reused as the
package set for the NixOS test system.

However, if we restore the original code removed by that commit:

```nix
nixpkgs.pkgs = lib.mkDefault pkgs;
```

… then that won't work because then we can't override `nixpkgs.system`
and reinstantiate that package set.  The whole *point* of this set of
changes is that we DON'T want to use the user's `pkgs` set as is
because then the wrong system is used for the NixOS test (on macOS).
Specifically, on Apple Silicon the `pkgs` in `pkgs.nixosTest` is going
to be instantiated with `system = aarch64-darwin;` but we want the
`nixpkgs.pkgs` that the NixOS machine uses to be that same package set
reinstantiated with `system = aarch64-linux;`.

Then I realized we can strike a balance by doing this instead:

```nix
nixpkgs.overlays = lib.mkDefault pkgs.overlays;
nixpkgs.config = lib.mkDefault pkgs.config;
```

… then we can get something which splits the difference: we preserve
the user's original package overrides but now we can reinstantiate
`nixpkgs.pkgs` with the new (correct) `nixpkgs.system`.

This isn't a foolproof solution, though, because `pkgs.overlays` and
`pkgs.config` do not necessarily contain all the information to
recreate the original `pkgs` that the user supplied.  As a concrete
example, if the user were to extend the package set "artificially" like
this:

```nix
(pkgs // { foo = …; }).nixosTest …
```

… then that `pkgs.foo` attribute would no longer be passed along to
`nixpgs.pkgs`.  However, in practice I believe most users are going to
be overriding `pkgs` using either the `config.packageOverrides` or
`overlays` arguments to the Nixpkgs function, so I think this is a
reasonable compromise.  Moreover, the benefit of this compromise is
that the NixOS test framework is much more "fire and forget" (we don't
need to instruct the users to instantiate Nixpkgs themselves twice for
the host and guest architecture).

tfc commented:
Just tested it. Works great on my aarch64 Macbook!
Tried to reproduce the mentioned problems on the Intel Macbook, but the macOS there is so old that Nix regularly breaks there (i guess that's because Nix reflects modern macOS), so i would abandon testing it there.

Gabriella439 commented:
@roberth: I think this is finally ready for a second look.  I believe I did this as closely as possible as to what you originally envisioned.  See the commit message in ab45d3118634ad3cd0c319e5bdb42e31524ecb0d which explains the balance that I had to strike to get this to work while breaking existing expectations/workflows as little as possible.

I can also clean up the commit history if desired (although I'm guessing this change is small enough to just squash merge it), but I'm leaving the commit history as-is for now so that the original thought process is preserved.

roberth committed:
nixosTest: Make similar to previous implementation again

I'd considered suggesting to add a conditional here, but that would
create VM-host-dependent portability issues for no good reason.

roberth committed:
pkgs.pkgsLinux: init

roberth committed:
nixosTest: Use memoized pkgsLinux

This solves the previously introduced evaluation performance
regression when evaluating multiple tests at once.

roberth committed:
testers.runNixOSTest: Make it work on non-linux VM hosts

roberth committed:
treewide: Make some NixOS tests darwin-host-proof

roberth committed:
nixos/all-tests: Make runTest work on darwin

roberth committed:
nixos/testing/pkgs: Warn darwin users about ambiguous `pkgs`

roberth commented on pkgs/build-support/testers/default.nix:
```diff
@@ -123,7 +123,18 @@
+              # `config.packageOverrides` arguments to Nixpkgs will be passed
+              # through to `nixpkgs.pkgs`.
+              config.nixpkgs.overlays = lib.mkDefault pkgs.overlays;
+              config.nixpkgs.config = lib.mkDefault pkgs.config;
```
This was too much of a breaking change, regressing evaluation performance and possibly other things, considering that this part is heavily, thoroughly used.

Fortunately, we can don't have to pick the same solution in both places. I'm sending a few commits that bring this closer to the previous solution and fix some other small problems.

Gabriella439 commented:
@roberth: Thanks!  Your changes look good to me

tfc commented:
Works great!

Gabriella439 commented:
I can go ahead and merge this once CI passes if nobody objects since I think all of @roberth and @tfc and I are satisfied

Gabriella439 commented:
Is it normal for `ofborg-eval` to take this long?  I'm not sure how I can monitor its progress

Gabriella439 commented:
@ofborg eval

cole-h commented on pkgs/top-level/stage.nix:
```diff
@@ -243,6 +243,16 @@ let
+    # If already linux: the same package set unaltered
+    # Otherwise, return a natively built linux package set for the current cpu architecture string.
+    # (ABI and other details will be set to the default for the cpu/os pair)
+    pkgsLinux =
```
I think this might be running into the same issue as when you were using `pkgsLocal` (https://github.com/NixOS/nixpkgs/pull/282401#discussion_r1461062860). ofborg might be happier if you exclude `pkgsLinux` from https://github.com/NixOS/nixpkgs/blob/1b439a273dadd4c675e81a0e05abbc9258889c02/pkgs/top-level/release-attrpaths-superset.nix

Gabriella439 committed:
Exclude `pkgsLinux`

… as suggested by @cole-h

Gabriella439 committed:
Merge branch 'master' into gabriella/macOS_NixOS_test

Gabriella439 commented:
Thanks to everybody for y'all's help!

414owen commented:
Hi @Gabriella439 ,

Thanks for working on this.
I've put your flake from the PR description into an empty git repo, and run `nix flake check`, making sure I'm on nix 2.19.2.

Doing so gives me:

```
$ nix flake check
warning: Git tree '/Users/owen/src/nixos-test-darwin' is dirty
warning: Ignoring setting 'auto-allocate-uids' because experimental feature 'auto-allocate-uids' is not enabled
warning: Ignoring setting 'impure-env' because experimental feature 'configurable-impure-env' is not enabled
error: a 'aarch64-linux' with features {} is required to build '/nix/store/nv3ziyvxfnyl4z0ayg9j25kvgp27ziyr-append-initrd-secrets.drv', but I am a 'aarch64-darwin' with features {apple-virt, benchmark, big-parallel, nixos-test}
```

I'm not sure how to go about debugging this. I can list the referrers, but it just tells me that there's a `nixos-system-machine-test` that needs it.

Is there a step I've missed? I got the impression that I don't need to specify pkgs in the machine configuration anymore? Is it something to do with `hvf/apple-virt`?

tfc commented:
@414owen you need a linux builder, see https://nixcademy.com/2024/02/12/macos-linux-builder/

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/macos-support-for-running-nixos-tests/40801/1

roberth commented:
Doc update, please review:

- https://github.com/NixOS/nixpkgs/pull/293573

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/nixostest-vs-runtest/26419/4
